src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,491 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Linq;
+
+namespace Microsoft.Collections.Extensions
+{
+    /// <summary>
+    /// DictionarySlim<TKey, TValue> is similar to Dictionary<TKey, TValue> but optimized in three ways:
+    /// 1) It allows access to the value by ref replacing the common TryGetValue and Add pattern.
+    /// 2) It does not store the hash code (assumes it is cheap to equate values).
+    /// 3) It does not accept an equality comparer (assumes Object.GetHashCode() and Object.Equals() or overridden implementation are cheap and sufficient).
+    /// </summary>
+    [DebuggerTypeProxy(typeof(Extensions.DictionarySlimDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class DictionarySlim<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey>
+    {
+        // We want to initialize without allocating arrays. We also want to avoid null checks.
+        // Array.Empty would give divide by zero in modulo operation. So we use static one element arrays.
+        // The first add will cause a resize replacing these with real arrays of three elements.
+        // Arrays are wrapped in a class to avoid being duplicated for each <TKey, TValue>
+        private static readonly Entry[] InitialEntries = new Entry[1];
+        // 1-based index into _entries; 0 means empty
+        private int[] _buckets;
+        private Entry[] _entries;
+        // 0-based index into _entries of head of free chain: -1 means empty
+        private int _freeList = -1;
+        private int _count;
+
+        [DebuggerDisplay("({key}, {value})->{next}")]
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            // 0-based index of next entry in chain: -1 means end of chain
+            // also encodes whether this entry _itself_ is part of the free list by changing sign and subtracting 3,
+            // so -2 means end of free list, -3 means index 0 but on free list, -4 means index 1 but on free list, etc.
+            public int next;
+        }
+
+        public DictionarySlim()
+        {
+            _buckets = HashHelpers.DictionarySlimSizeOneIntArray;
+            _entries = InitialEntries;
+        }
+
+        public DictionarySlim(int capacity)
+        {
+            capacity = HashHelpers.GetPrime(capacity);
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+        }
+
+        public int Count => _count;
+
+        public int Capacity => _entries.Length;
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
```

Comment:
Should there be a not null check so as to not result in a `NullReferenceException` or since this is focused on performance is a `NullReferenceException` acceptable and may eventually be mitigated by C# 8.0's Nullable Reference types by annotating the type as non-nullable?

Reply:
No, that's an oversight, all parameters should be checked. We might possibly want to use `ThrowHelper` if it affects perf.